src/main/java/io/temporal/internal/replay/ReplayWorkflowTaskHandler.java

```diff
@@ -104,10 +108,27 @@ public ReplayWorkflowTaskHandler(
       return handleWorkflowTaskImpl(workflowTask.toBuilder(), metricsScope);
     } catch (Throwable e) {
       metricsScope.counter(MetricsType.WORKFLOW_TASK_EXECUTION_FAILURE_COUNTER).inc(1);
+      // Fail workflow and not a task as WorkflowExecutionException is thrown only if FailWorkflow
+      // policy was set.
+      if (e instanceof WorkflowExecutionException) {
+        RespondWorkflowTaskCompletedRequest response =
+            RespondWorkflowTaskCompletedRequest.newBuilder()
+                .setTaskToken(workflowTask.getTaskToken())
+                .setIdentity(options.getIdentity())
+                .addCommands(
+                    Command.newBuilder()
+                        .setCommandType(CommandType.COMMAND_TYPE_FAIL_WORKFLOW_EXECUTION)
+                        .setFailWorkflowExecutionCommandAttributes(
+                            FailWorkflowExecutionCommandAttributes.newBuilder()
+                                .setFailure(((WorkflowExecutionException) e).getFailure()))
+                        .build())
+                .build();
+        return new WorkflowTaskHandler.Result(workflowType, response, null, null, null, false);
+      }
       // Only fail workflow task on the first attempt, subsequent failures of the same workflow task
       // should timeout. This is to avoid spin on the failed workflow task as the service doesn't
       // yet increase the retry interval.
-      if (workflowTask.getAttempt() > 0) {
+      if (workflowTask.getAttempt() > 1) {
```

Comment:
hopefully we added a unit test to cover this as this should have been caught when I was syncing to Alex's latest proto changes

Reply:
Good catch. We had a test that ensured that the worker doesn't spin on the task failure. But after I added an additional check that the task failure is reported exactly once I found that it wasn't doing so. [Here is](https://github.com/temporalio/java-sdk/pull/172/commits/250f70dc022bffc0c2ea350a082fb8b46e6067ec) the new unit test and the fix. The unit test fails with attempts condition changed.